blocks/main_loop.py

```diff
@@ -156,16 +159,17 @@ def _check_finish_training(self):
         # the iteration the corresponding log record can be found only in
         # the previous row.
         if (self.log.current_row.training_finish_requested or
-                self.log.current_row.keyboard_interrupt_received or
-                self.log.previous_row.keyboard_interrupt_received):
+                self.log.current_row.interrupt_received or
+                self.log.previous_row.interrupt_received):
             raise TrainingFinish
 
-    def _handle_keyboard_interrupt(self, signal_number, frame):
+    def _handle_interrupt(self, signal_number, frame):
         # After receiving a first keyboard interrupt signal,
         # ignore all following ones.
```

Comment:
Not exactly, because the `finally` clause is still run (https://github.com/bartvm/blocks/blob/interrupt/blocks/main_loop.py#L101), and I think after the second interrupt it shouldn't. And in case of the cluster we might need different behaviour I guess: try hard to finish gracefully until killed blatantly by SIGKILL.